src/text/parsers/numeric_support.rs

```diff
@@ -0,0 +1,97 @@
+//! This module contains parsing functions that are useful in parsing multiple numeric Ion types.
+
+use nom::{IResult, Parser};
+use nom::branch::alt;
+use nom::bytes::streaming::tag;
+use nom::combinator::{recognize, opt, map};
+use nom::sequence::{pair, terminated, preceded};
+use nom::character::streaming::{char, one_of, digit1};
+use nom::multi::many0_count;
+
+/// Recognizes the digits of a base-10 integer. (i.e. An integer without a sign.)
+pub(crate) fn base_10_integer_digits(input: &str) -> IResult<&str, &str> {
+    alt((
+        // The number is either a zero...
+        recognize(char('0')),
+        // Or it's a non-zero followed by some number of '_'-separated digits
+        digits_before_dot
+    ))(input)
+}
+
+/// Recognizes either:
+/// * a zero
+/// * a non-zero followed by some number of digits with optional underscores
+pub(crate) fn digits_before_dot(input: &str) -> IResult<&str, &str> {
+    alt((
+        tag("0"),
+        recognize(
+            pair(leading_digit, trailing_digits)
+        )
+    ))(input)
+}
+
+/// Recognizes the first digit of a multi-digit base-10 integer. (i.e. Any digit but zero.)
```

Comment:
```suggestion
/// Recognizes the first non-0 digit of a multi-digit base-10 integer. (i.e. Any digit but zero.)
```

Reply:
I'd prefer to leave this as-is. In my opinion the rephrasing suggests that this will look at several digits and match the first non-zero one rather than matching the first digit, which must be non-zero.